functions/executor/call.py

```diff
@@ -209,40 +267,52 @@ def _exec_dag_function_causal(pusher_cache, kvs, triggers, function, schedule):
+                new_trigger.versioned_key_locations[addr].versioned_keys.extend(
+                                    versioned_key_locations[addr].versioned_keys)
 
             for key in dependencies:
+                #logging.info("to send trigger dependency includes key %s" % key)
                 dep = new_trigger.dependencies.add()
                 dep.key = key
-                dep.vector_clock = dependencies[key]
+                dep.vector_clock.update(dependencies[key])
 
             dest_ip = schedule.locations[conn.sink]
             sckt = pusher_cache.get(sutils._get_dag_trigger_address(dest_ip))
             sckt.send(new_trigger.SerializeToString())
 
     if is_sink:
+        #logging.info('DAG %s (ID %s) completed in causal mode; result at %s.' %
+        #        (schedule.dag.name, schedule.id, schedule.output_key))
+
         vector_clock = {}
-        if schedule.response_id in dependencies:
-            if schedule.id in dependencies[schedule.response_id]:
-                dependencies[schedule.response_id][schedule.id] += 1
+        if schedule.output_key in dependencies:
+            if schedule.client_id in dependencies[schedule.output_key]:
+                dependencies[schedule.output_key][schedule.client_id] += 1
             else:
-                dependencies[schedule.response_id][schedule.id] = 1
-            vector_clock = dependencies[schedule.response_id][schedule.id]
-            del dependencies[schedule.response_id]
+                dependencies[schedule.output_key][schedule.client_id] = 1
+            vector_clock.update(dependencies[schedule.output_key])
+            del dependencies[schedule.output_key]
         else:
-            vector_clock = {schedule.id: 1}
+            vector_clock = {schedule.client_id : 1}
 
-        succeed = kvs.causal_put(schedule.response_id,
+        succeed = kvs.causal_put(schedule.output_key,
                                  vector_clock, dependencies,
-                                 serialize_val(result), schedule.id)
+                                 serialize_val(result), schedule.client_id)
         while not succeed:
-            kvs.causal_put(schedule.response_id, vector_clock,
-                           dependencies, serialize_val(result), schedule.id)
+            kvs.causal_put(schedule.output_key, vector_clock,
+                           dependencies, serialize_val(result), schedule.client_id)
 
+        # issue requests to GC the version store
+        for cache_addr in versioned_key_locations:
+            gc_addr = cache_addr[:-4] + str(int(cache_addr[-4:]) - 50)
+            #logging.info("cache GC addr is %s" % gc_addr)
+            sckt = pusher_cache.get(gc_addr)
+            sckt.send_string(schedule.client_id)
 
 def _exec_func_causal(kvs, func, args, kv_pairs,
-                      schedule, versioned_key_locations):
-    func_args = ()
+                      schedule, versioned_key_locations, dependencies):
+    func_args = []
```

Comment:
I remember it's because tuple doesn't support item assignment, so `func_args[key_index_map[key]] = xxx` will throw an error.
But list is fine, so I first define it as a list and later on when the arguments are populated I convert it back to a tuple